model/src/main/kotlin/ScanSummary.kt

```diff
@@ -59,6 +59,13 @@ data class ScanSummary(
     @JsonAlias("fileCount")
     val fileCount: Int,
 
+    /**
+     * The [SPDX package verification code](https://spdx.org/spdx_specification_2_0_html#h.2p2csry), calculated from
+     * all files in the package. Note that if the scanner is configured to ignore certain files they will still be
+     * included in the calculation of this code.
```

Comment:
I'm curious what's the rationale for doing it this way. After all, this is the *scan* summary, so I'd expect all meta-data to refer to what was *scanned*.

Also, if this would only refer to the scanned files, we could probably apply some performance optimizations, like not calculating the individual file SHA1s or even the whole package verification code our self (again), but taking it from the scanner's raw result, if these fields are present.

Reply:
I'm not yet 100% convinced as the linked SPDX docs say "This identifier enables a recipient to determine if any file in the original package (that the analysis was done on)". What concerns me a bit is the "the analysis was done on" part, where *analysis* in our case is the *scan*. To me this sentence is a bit contradicting because "original package" and "the analysis was done on" are not necessarily the same thing, esp. as the SPDX docs themselves say that e.g. the SPDX file itself should be excluded.

Reply:
IMO the phrase "that the analysis was done on" does not imply that the analysis has to cover all files, so it does not contradict including all files in the code even if some were not scanned.

The documentation also says that the SPDX analysis file has to be excluded from the verification code, but currently we include all files. As the SPDX file is called "[packageName].spdx" we can hardly determine how exactly the file will be called, so the only option I see is to exclude all files with ".spdx" extension, but that might also be wrong.

@tsteenbe What is your opinion on that?

Reply:
Let me double check with @kestewart  @goneall

Reply:
If I recall the discussions when we devised the verification code algorithm correctly, the intent of the excluded files was to allow any tool which generates a verification code to exclude files that were either a) impossible to calculate or b) were not considered part of the analyzed files from the tool's perspective.

For a), one scenario discussed was if the tool were to generate the SPDX file which is included in the package then it was impossible to know the checksum prior to creation of the document itself and the document contains the verification code being calculated.  I believe that is one of the main reasons for excluding the SPDX file.  I don't recall any discussion of embedded or other SPDX documents.

I don't recall much discussion on b).

We created an excluded files list which must be included along with the verification code to specify which files were excluded (see the spec section 3.9.6).

As long as "important files" are included and any excluded files are listed in the excluded files list  then the calculations can be replicated and the file verified.

For the SPDX files excluded from the analysis, IMO just exclude the one file that is associated with the package for which the verification code is being generated.

Reply:
Thanks for the feedback @goneall. Because at this time we mainly use this for third-party libraries I will adapt this change to exclude all ".spdx" files found in the repository and properly list them in the code.

Reply:
In my view there's still the open question whether @goneall's mentioned *b) case* in our situation applies to "which files have been scanned by the scanner". My current feeling is yes, it does, and we should not only exclude ".spdx" files from package verification code calculation, but also [all files that we ignore while scanning](https://github.com/heremaps/oss-review-toolkit/blob/2edd9c01239e4949294ec9518314c4378a9d71c2/scanner/src/main/kotlin/scanners/ScanCode.kt#L99-L103), namely `ORT_CONFIG_FILENAME` and `NON_LICENSE_FILENAMES`.

Reply:
According to @goneall the reason for having the excludes list is that the SPDX file should contain the verification code and can therefore not be part of its calculation. As the scanner does not modify any files in the repository I do not really see a reason for excluding any files from this perspective. Excluding "*.sdpx" makes it at least more likely that our code matches the code contained in the projects SPDX file.
To me the value of the code is more to identify the whole source set, not only the scanned files. E.g. if the the sources are scanned with another scanner or another excludes configuration, was it still exactly the same source code which was scanned? This check is not possible anymore if we exclude files based on the scanner configuration. Also I'm a bit concerned that the file list can become very long if many files are excluded from the scan and that this was not the intention when the exclude list was added to the specs.

Reply:
> According to @goneall the reason for having the excludes list is that the SPDX file should contain the verification code and can therefore not be part of its calculation.

But also according to @goneall, this is just *one* example, and not the *only* reason.

> To me the value of the code is more to identify the whole source set, not only the scanned files.

Yeah, this somewhat semi-convinced me. Actually scanners might have built-in excludes (like ScanCode [ignores](https://github.com/nexB/scancode-toolkit/blob/8788fa3408fedb8a0d4532c1025dab114462c9c9/src/commoncode/ignore.py#L81) e.g. a `.git` directory) that you can't modify, so we don't even necessarily know beforehand which files a scanner would scan unless we look at the result afterwards. With this in mind, it does not make sense to try to match the set of files considered for the package verification code with the set of files scanned by the scanner.

But this brings the topic of VCS directories like `.git` on the table: We *should* ignore those when calculating the package verification code, right?

Reply:
> But this brings the topic of VCS directories like `.git` on the table: We _should_ ignore those when calculating the package verification code, right?

Yes, that makes a lot of sense. But I wonder if they should still be listed in the excludes list of the vericiation code, I have a tendency towards no, because it would blow up the code and the files can be seen as metadata that is not really part of the source code.
@goneall @tsteenbe have you discussed earlier how such files should be handled?

I would add a list of default excludes for the VCS supported by ORT: ".git", ".hg", ".svn", ".cvs", ".repo"
Do you want to add any others to the list?

Reply:
Regarding listing of the excludes: Is there a convention that entries with a trailing slash denote excluded directories? Then we could add ".git/" to the list instead of listing all contents of the ".git" directory individually.

For CVS the directory name would actually be "CVS" (not ".cvs"), but other than that I guess that list is complete: It's probably enough for now to only ignore directories for those VCS that the ORT Downloader itself supports.

Reply:
Done, without including them in the list.